### PR TITLE
DOXIA-569 add Markdown output (sink)

### PR DESCRIPTION
**Pull request by hboutemy**
https://issues.apache.org/jira/browse/DOXIA-569

### CODE REVIEW

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ */
+@SuppressWarnings( "checkstyle:interfaceistype" )
+public interface MarkdownMarkup
+    extends TextMarkup
+{
+    // ----------------------------------------------------------------------
+    // Markup separators
+    // ----------------------------------------------------------------------
+
+    /** APT backslash markup char: '\\' */
```
Comment (kwin) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:36]: why do we need APT constants in this class?

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 2]
```diff
@@ -0,0 +1,1180 @@
...
+     * Escape special characters in a text in APT:
+     *
+     * <pre>
+     * \~, \=, \-, \+, \*, \[, \], \<, \>, \{, \}, \\
+     * </pre>
+     *
+     * @param text the String to escape, may be null
+     * @return the text escaped, "" if null String input
+     */
+    private static String escapeAPT( String text )
```
Comment (kwin) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:1143]: why APT in MD sink?

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java (hunk 1 above)
Comment (hboutemy) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:36]: as you can see, I created the sink by copy/pasting APT sink and updating content: yes, I did not change each and every comment

File: doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java [hunk 3]
```diff
@@ -492,7 +492,7 @@ public void testFooter()
      * the same result as {@link #getListBlock getListBlock}( item ).
      *
      */
-    public void testList()
+    public void _testList()
     {
```
Comment (michael-o) [inline doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java:496]: Why this change?

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java (hunk 1 above)
Comment (michael-o) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:36]: Still should be fixed.

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    // ----------------------------------------------------------------------
+    // Markup syntax
+    // ----------------------------------------------------------------------
+
+    /** Syntax for the anchor end: "\"></a>" */
+    String ANCHOR_END_MARKUP = "\"></a>";
+
+    /** Syntax for the anchor start: "<a name=\"" */
+    String ANCHOR_START_MARKUP = "<a name=\"";
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:74]: Markdown will never use `name` for anchors. Only `id`.

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 5]
```diff
@@ -0,0 +1,1163 @@
...
+        {
+            buffer.append( EOL );
+        }
+        else if ( verbatimFlag )
+        {
+            write( EOL );
+        }
+        else
+        {
+            write( "\\" + EOL );
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:994]: There is a final static for backslash.

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java (hunk 2 above)
Comment (michael-o) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:1127]: Must be fixed as well

**Review by michael-o — CHANGES_REQUESTED** [review]
Please note when this is going to be merged into master some things might have changed in terms of conversion due to behavioral changes in Doxia 2.x.

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java (hunk 1 above)
Comment (hboutemy) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java]: done

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java (hunk 2 above)
Comment (hboutemy) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java]: done

File: doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java (hunk 3 above)
Comment (hboutemy) [reply doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java]: test disabled temporarly: now that it works, we can re-enable it again, good catch, thank you
re-enabled

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java (hunk 4 above)
Comment (hboutemy) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:68]: currently, code is in Doxia 1, which is still using name, isn't it?
you're right that id will be required when merging to Doxia 2

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java (hunk 5 above)
Comment (hboutemy) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java]: done

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java (hunk 4 above)
Comment (michael-o) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:68]: Yes, correct. You need to keep an eye on the changes from 1 to 2.

Comment (hboutemy) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:68]: I'll let you merge from 1 to 2 :)
I don't think there are many changes

Comment (michael-o) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownMarkup.java:68]: Grrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr....

**Comment by hboutemy** [comment]
merged in Doxia 1.x in b174a89b270a12e5871ddaf97c3e39594f284c79

**Comment by hboutemy** [comment]
merged in master in df9353869e6031987533ec2f288eede1f945728a
